### PR TITLE
[BOLT] Add structure of CDSplit to SplitFunctions

### PR DESCRIPTION
**Pull request by ShatianWang**
This commit establishes the general structure of the CDSplit strategy in SplitFunctions without incorporating the exact splitting logic. With -split-functions -split-strategy=cdsplit, the SplitFunctions pass will run twice: the first time is before function reordering and functions are hot-cold split; the second time is after function reordering and functions are hot-warm-cold split based on the fixed function ordering. Currently, all functions are hot-warm split after the entry block in the second splitting pass. Subsequent commits will introduce the precise splitting logic. NFC.

### CODE REVIEW

**Comment by ShatianWang** [comment]
A test will be added in a later PR with the complete CDSplit logic.

File: bolt/lib/Passes/SplitFunctions.cpp [hunk 1]
```diff
@@ -409,8 +487,10 @@ void SplitFunctions::splitFunction(BinaryFunction &BF, SplitStrategy &S) {
     LLVM_DEBUG(dbgs() << "Estimated size for function " << BF
                       << " post-split is <0x" << Twine::utohexstr(HotSize)
                       << ", 0x" << Twine::utohexstr(ColdSize) << ">\n");
-    if (alignTo(OriginalHotSize, opts::SplitAlignThreshold) <=
-        alignTo(HotSize, opts::SplitAlignThreshold) + opts::SplitThreshold) {
+    if (S.autoReversal() &&
```
Comment (maksfb) [inline bolt/lib/Passes/SplitFunctions.cpp:490]: Do you mean to run `calculateEmittedSize()` above even when `autoReversal()` is off?

Comment (ShatianWang) [reply bolt/lib/Passes/SplitFunctions.cpp]: Removed `autoReversal()` per our offline discussion.

**Comment by ShatianWang** [comment]
Based on comments from and offline discussion with @maksfb, added [fixup! [BOLT] Add structure of CDSplit to SplitFunctions](https://github.com/llvm/llvm-project/pull/73430/commits/165b75fbeb81af70629007838f76f784eec712d5)

File: bolt/lib/Rewrite/BinaryPassManager.cpp [hunk 2]
```diff
@@ -430,6 +430,8 @@ void BinaryFunctionPassManager::runAllPasses(BinaryContext &BC) {
   Manager.registerPass(
       std::make_unique<ReorderFunctions>(PrintReorderedFunctions));
 
+  Manager.registerPass(std::make_unique<SplitFunctions>(PrintSplit));
```
Comment (maksfb) [inline bolt/lib/Rewrite/BinaryPassManager.cpp:433]: Please add a brief comment why we need another pass of SplitFunctions here too.

File: bolt/lib/Passes/SplitFunctions.cpp [hunk 3]
```diff
@@ -136,6 +149,55 @@ struct SplitProfile2 final : public SplitStrategy {
...
+  BinaryContext &BC;
+  using BasicBlockOrder = BinaryFunction::BasicBlockOrderType;
+
+  explicit SplitCacheDirected(BinaryContext &BC) : BC(BC) {}
+
+  bool canSplit(const BinaryFunction &BF) override {
+    return BF.hasValidProfile() && hasFullProfile(BF) && !allBlocksCold(BF);
+  }
+
+  bool keepEmpty() override { return true; }
```
[the remark this answers is not in the document]
Comment (maksfb) [reply bolt/lib/Passes/SplitFunctions.cpp:152]: `bool compactFragments() override { return false; }`?

**Comment by ShatianWang** [comment]
New fixup addressing the new comments by @maksfb .

**Review by maksfb — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**